scripts/train/train.py

```diff
@@ -164,170 +164,264 @@ def build_dataloader(cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
+    # TODO remove this as we should treat configs as immutable objects
     cfg = update_batch_size_info(cfg)
 
-    # Read FSDP Config as a dict
-    fsdp_config = cfg.get('fsdp_config', None)
-    fsdp_config = om.to_container(fsdp_config,
-                                  resolve=True) if fsdp_config else None
-    assert isinstance(fsdp_config, Dict) or fsdp_config is None
+    print('Logging config...')
+    log_config(cfg)
+
+    # Mandatory model training configs
+    model_config: DictConfig = pop_config(cfg, 'model', must_exist=True)
+    tokenizer_config: DictConfig = pop_config(cfg, 'tokenizer', must_exist=True)
+    optimizer_config: DictConfig = pop_config(cfg, 'optimizer', must_exist=True)
+    scheduler_config: DictConfig = pop_config(cfg, 'scheduler', must_exist=True)
+    train_loader_config: DictConfig = pop_config(cfg,
+                                                 'train_loader',
+                                                 must_exist=True)
+
+    # Optional fsdp data, fine-tuning, and eval configs
+    fsdp_config: Optional[DictConfig] = pop_config(cfg,
+                                                   'fsdp_config',
+                                                   must_exist=False)
+    lora_config: Optional[DictConfig] = pop_config(cfg,
+                                                   'lora',
+                                                   must_exist=False)
+    eval_loader_config: Optional[DictConfig] = pop_config(cfg,
+                                                          'eval_loader',
+                                                          must_exist=False)
+    icl_tasks_config: Optional[DictConfig] = pop_config(cfg,
+                                                        'icl_tasks',
+                                                        must_exist=False)
+
+    # Optional logging, evaluation and callback configs
+    logger_configs: Optional[DictConfig] = pop_config(cfg,
+                                                      'loggers',
+                                                      must_exist=False)
+    callback_configs: Optional[DictConfig] = pop_config(cfg,
+                                                        'callbacks',
+                                                        must_exist=False)
+    algorithm_configs: Optional[DictConfig] = pop_config(cfg,
+                                                         'algorithms',
+                                                         must_exist=False)
+
+    # Mandatory parameters for trainer
+    device_train_batch_size: int = cfg.pop('device_train_batch_size')
+    device_eval_batch_size: int = cfg.pop('device_eval_batch_size')
+    max_duration: Union[int, str] = cfg.pop('max_duration')
+    eval_interval: Union[int, str] = cfg.pop('eval_interval')
+    precision: str = cfg.pop('precision')
+    max_seq_len: int = cfg.pop('max_seq_len')
+
+    # Optional parameters for will be set to default values if not specified.
```

Comment:
Note to self to double check all the defaults

Reply:
checked